File: JavaCode/ModelException.java
package com.ndepend.inspections;

public class ModelException extends Exception {
    public ModelException(XmlElement xmlElement, String[] tagsExpected) {
        super("Expected "+Aggregate(tagsExpected)+" tag but xml element name was "+xmlElement.Name);
    }

    public ModelException(XmlAttribute xmlAttribute, String[] tagsExpected) {
        super("Expected "+Aggregate(tagsExpected)+" tag but xml attribute name was "+xmlAttribute.Name);
    }

    public ModelException(String message) {
        super(message);
    }

    private static String Aggregate(String[] arr) {
        StringBuilder sb = new StringBuilder(arr[0]);
        for (int i = 1; i < arr.length; i++) {
            sb.append(" or ");
            sb.append(arr[i]);
        }
        return sb.toString();
    }
}

File: SimpleXmlParser/InspectionsModel/ModelException.cs

using System.Text;

namespace SimpleXmlParser.InspectionsModel {
   public class ModelException : Exception {
      internal ModelException(XmlElement xmlElement, params string[] tagsExpected) : base(
         $"Expected {Aggregate(tagsExpected, " or ")} tag but xml element name was {xmlElement.Name}") { }

      internal ModelException(XmlAttribute xmlAttribute, params string[] tagsExpected) : base(
         $"Expected {Aggregate(tagsExpected, " or ")} tag but xml attribute name was {xmlAttribute.Name}") { }

      internal ModelException(string msg) : base(msg) { }


      private static string Aggregate(string[] arr, string sep) {
         StringBuilder sb = new StringBuilder(arr[0]);
         for (int i = 1; i < arr.Length; i++) {
            sb.Append(sep);
            sb.Append(arr[i]);
         }
         return sb.ToString();
      }
   }
}
